File: src/services/reporting/batchService.js
import { BehaviorSubject } from '../behaviorSubject.js';
import { authService } from '../authService.js';

class BatchService {
  constructor() {
    this.activeBatch = new BehaviorSubject(null);
    this.currentBatchStats = new BehaviorSubject(null);

    this._db = null;
    this._midnightCheckTimer = null;
    this._midnightDailyTimer = null;
  }

  async init(db) {
    this._db = db;
    await this.initializeBatchSystem();
    this._startMidnightCheckTimer();
    return this;
  }

  dispose() {
    clearTimeout(this._midnightCheckTimer);
    clearInterval(this._midnightDailyTimer);
  }

  async initializeBatchSystem() {
    try {
      const locationId = authService.selectedLocation.value?.id;
      if (!locationId) return;

      const existingBatch = await this._db.batchReports.getActiveBatchReport(String(locationId));

      if (existingBatch) {
        this.activeBatch.add(existingBatch);

        if (this._shouldAutoCloseBatch(existingBatch.start_date)) {
          await this._autoCloseBatch(existingBatch);
          await this._createNewBatch();
        }
      } else {
        await this._createNewBatch();
      }
    } catch (e) {
      console.error('[BatchService] Error initializing batch system:', e);
    }
  }

  async _createNewBatch() {
    try {
      const locationId = authService.selectedLocation.value?.id;
      const terminalId = authService.selectedTerminal.value?.id;
      if (!locationId || !terminalId) return false;

      const uuid = crypto.randomUUID();
      const now = new Date();
      const batchName = `Batch ${this._formatDate(now)}`;

      const newBatch = {
        uuid,
        batch_name: batchName,
        start_date: now.toISOString(),
        status: 'ACTIVE',
        location_id: String(locationId),
        terminal_id: String(terminalId),
        total_sales: 0.0,
        total_cost: 0.0,
        gross_profit: 0.0,
        net_profit: 0.0,
        total_orders: 0,
        total_items_sold: 0,
        created_at: now.toISOString(),
        updated_at: now.toISOString(),
      };

      await this._db.batchReports.createBatchReport(newBatch);
      this.activeBatch.add(newBatch);
      return true;
    } catch (e) {
      console.error('[BatchService] Error creating new batch:', e);
      return false;
    }
  }

  async startNewBatch() {
    try {
      if (this.activeBatch.value?.status === 'ACTIVE') {
        const closed = await this.closeBatch(this.activeBatch.value.uuid);
        if (!closed) return false;
      }
      return await this._createNewBatch();
    } catch (e) {
      console.error('[BatchService] Error starting new batch:', e);
      return false;
    }
  }

  async closeBatch(uuid) {
    try {
      const stats = await this.calculateBatchStats(uuid);
      if (!stats) return false;

      const totalItems = stats.topProducts.reduce((sum, p) => sum + p.quantitySold, 0);

      const success = await this._db.batchReports.closeBatchReport(uuid, {
        totalSales: stats.grossSales,
        totalCost: 0.0,
        grossProfit: stats.grossSales,
        netProfit: stats.netSales,
        totalOrders: stats.totalOrders,
        totalItemsSold: totalItems,
        categoryBreakdown: this._convertCategoryStatsToMap(stats.categoryStats),
        productBreakdown: this._convertTopProductsToMap(stats.topProducts),
        paymentMethodBreakdown: stats.paymentMethodBreakdown,
        serviceTypeBreakdown: stats.serviceTypeBreakdown,
      });

      if (success) {
        const current = this.activeBatch.value;
        if (current) {
          this.activeBatch.add({ ...current, status: 'CLOSED', end_date: new Date().toISOString(), closed_at: new Date().toISOString() });
        }
      }

      return success;
    } catch (e) {
      console.error('[BatchService] Error closing batch:', e);
      return false;
    }
  }

  async _autoCloseBatch(batch) {
    try {
      await this.closeBatch(batch.uuid);
    } catch (e) {
      console.error('[BatchService] Error auto-closing batch:', e);
    }
  }

  _shouldAutoCloseBatch(batchStartDate) {
    const start = new Date(batchStartDate);
    const now = new Date();
    return start.getFullYear() !== now.getFullYear() ||
      start.getMonth() !== now.getMonth() ||
      start.getDate() !== now.getDate();
  }

  _startMidnightCheckTimer() {
    const now = new Date();
    const tomorrow = new Date(now.getFullYear(), now.getMonth(), now.getDate() + 1);
    const msUntilMidnight = tomorrow.getTime() - now.getTime();

    this._midnightCheckTimer = setTimeout(async () => {
      await this._onMidnightReached();
      this._midnightDailyTimer = setInterval(() => this._onMidnightReached(), 24 * 60 * 60 * 1000);
    }, msUntilMidnight);
  }

  async _onMidnightReached() {
    try {
      if (this.activeBatch.value?.status === 'ACTIVE') {
        await this._autoCloseBatch(this.activeBatch.value);
        await this._createNewBatch();
      }
    } catch (e) {
      console.error('[BatchService] Error handling midnight batch closure:', e);
    }
  }

  async calculateBatchStats(batchUuid) {
    try {
      const batch = await this._db.batchReports.getBatchReportByUuid(batchUuid);
      if (!batch) return null;

      const allOrders = await this._db.extOrders.getAllOrders();

      const batchStart = new Date(batch.start_date ?? Date.now());
      const batchEnd = batch.end_date ? new Date(batch.end_date) : null;

      const batchOrders = allOrders.filter(order => {
        const orderDate = new Date(order.created_at ?? Date.now());
        return orderDate > batchStart && (batch.status === 'ACTIVE' || !batchEnd || orderDate < batchEnd);
      });

      let grossSales = 0, discounts = 0, refunds = 0;
      let totalOrders = batchOrders.length;
      let completedOrders = 0, refundedOrders = 0, voidedOrders = 0, canceledOrders = 0;

      const paymentMethodBreakdown = {};
      const serviceTypeBreakdown = {};
      const hourlyOrderCount = {};
      const hourlySales = {};
      const categoryStats = {};
      const productDataMap = {};

      for (const order of batchOrders) {
        let orderTotal = 0;

        if (order.products) {
          for (const product of order.products) {
            const qty = this._parseDouble(product.qty ?? 1.0);
            const price = this._parseDouble(product.selling_price ?? 0.0);
            orderTotal += price * qty;

            const productUuid = product.uuid?.toString() ?? '';
            const productName = product.name ?? 'Unknown';
            const categoryUuid = product.category_uuid?.toString();

            if (productUuid) {
              if (productDataMap[productUuid]) {
                productDataMap[productUuid].quantitySold += qty;
                productDataMap[productUuid].revenue += price * qty;
              } else {
                productDataMap[productUuid] = {
                  productUuid, productName, quantitySold: qty, revenue: price * qty, categoryUuid,
                };
              }
            }

            if (categoryUuid) {
              if (categoryStats[categoryUuid]) {
                categoryStats[categoryUuid].itemsSold += qty;
                categoryStats[categoryUuid].revenue += price * qty;
                categoryStats[categoryUuid].orderCount += 1;
              } else {
                const categoryName = await this._getCategoryName(categoryUuid);
                categoryStats[categoryUuid] = {
                  categoryUuid, categoryName, itemsSold: qty, revenue: price * qty, orderCount: 1,
                };
              }
            }
          }
        }

        if (order.discount) {
          const discountValue = this._parseDiscount(order.discount, orderTotal);
          discounts += discountValue;
          orderTotal -= discountValue;
        }

        grossSales += orderTotal;

        const status = (order.last_status ?? '').toLowerCase();
        if (status === 'completed') completedOrders++;
        else if (status === 'refund') { refundedOrders++; refunds += orderTotal; }
        else if (status === 'voided') voidedOrders++;
        else if (status === 'canceled') canceledOrders++;

        const serviceType = order.service_type ?? 'Unknown';
        serviceTypeBreakdown[serviceType] = (serviceTypeBreakdown[serviceType] ?? 0) + orderTotal;

        if (order.payments) {
          for (const payment of order.payments) {
            const method = payment.type?.toString() ?? 'Unknown';
            const amount = this._parseDouble(payment.amount ?? 0);
            paymentMethodBreakdown[method] = (paymentMethodBreakdown[method] ?? 0) + amount;
          }
        }

        const orderDate = new Date(order.created_at ?? Date.now());
        const hour = `${String(orderDate.getHours()).padStart(2, '0')}:00`;
        hourlyOrderCount[hour] = (hourlyOrderCount[hour] ?? 0) + 1;
        hourlySales[hour] = (hourlySales[hour] ?? 0) + orderTotal;
      }

      const topProducts = Object.values(productDataMap).sort((a, b) => b.revenue - a.revenue).slice(0, 20);
      const netSales = grossSales - discounts - refunds;
      const averageOrderValue = totalOrders > 0 ? grossSales / totalOrders : 0;

      return {
        batchUuid, grossSales, netSales, discounts, refunds, tax: 0,
        totalOrders, completedOrders, refundedOrders, voidedOrders, canceledOrders,
        averageOrderValue, paymentMethodBreakdown, serviceTypeBreakdown,
        hourlyOrderCount, hourlySales, topProducts, categoryStats,
        calculatedAt: new Date().toISOString(),
      };
    } catch (e) {
      console.error('[BatchService] Error calculating batch stats:', e);
      return null;
    }
  }

  async updateBatchStats() {
    try {
      if (!this.activeBatch.value || this.activeBatch.value.status !== 'ACTIVE') return;

      const stats = await this.calculateBatchStats(this.activeBatch.value.uuid);
      if (!stats) return;

      this.currentBatchStats.add(stats);
      const totalItems = stats.topProducts.reduce((sum, p) => sum + p.quantitySold, 0);

      await this._db.batchReports.updateBatchReportStats(this.activeBatch.value.uuid, {
        totalSales: stats.grossSales,
        totalCost: 0,
        grossProfit: stats.grossSales,
        netProfit: stats.netSales,
        totalOrders: stats.totalOrders,
        totalItemsSold: totalItems,
      });
    } catch (e) {
      console.error('[BatchService] Error updating batch stats:', e);
    }
  }

  async getBatchHistory({ startDate, endDate, locationId } = {}) {
    try {
      const locId = locationId ?? String(authService.selectedLocation.value?.id ?? '');
      const start = startDate ?? new Date(Date.now() - 30 * 24 * 60 * 60 * 1000);
      const end = endDate ?? new Date();
      return await this._db.batchReports.getBatchReportsByDateRange({ startDate: start, endDate: end, locationId: locId });
    } catch (e) {
      console.error('[BatchService] Error getting batch history:', e);
      return [];
    }
  }

  async getBatchByUuid(uuid) {
    try {
      return await this._db.batchReports.getBatchReportByUuid(uuid);
    } catch (e) {
      console.error('[BatchService] Error getting batch by UUID:', e);
      return null;
    }
  }

  _parseDouble(value) {
    if (value == null) return 0;
    if (typeof value === 'number') return value;
    if (typeof value === 'string') return parseFloat(value) || 0;
    return 0;
  }

  _parseDiscount(discount, total) {
    try {
      if (!discount) return 0;
      const parts = discount.split(',');
      if (parts.length !== 2) return 0;
      const value = parseFloat(parts[0]) || 0;
      const type = parts[1].toLowerCase();
      if (type === 'percentage') return (total * value) / 100;
      if (type === 'fixed') return Math.min(value, total);
      return 0;
    } catch (_) { return 0; }
  }

  async _getCategoryName(categoryUuid) {
    try {
      return await this._db.categories.getCategoryNameByUuid(categoryUuid) ?? 'Unknown';
    } catch (_) { return 'Unknown'; }
  }

  _formatDate(date) {
    const y = date.getFullYear();
    const m = String(date.getMonth() + 1).padStart(2, '0');
    const d = String(date.getDate()).padStart(2, '0');
    return `${y}-${m}-${d}`;
  }

  _convertCategoryStatsToMap(categoryStats) {
    const result = {};
    for (const [key, val] of Object.entries(categoryStats)) {
      result[key] = { name: val.categoryName, items_sold: val.itemsSold, revenue: val.revenue, order_count: val.orderCount };
    }
    return result;
  }

  _convertTopProductsToMap(products) {
    const result = {};
    products.forEach((p, i) => {
      result[p.productUuid] = { name: p.productName, quantity_sold: p.quantitySold, revenue: p.revenue, rank: i + 1 };
    });
    return result;
  }
}

export const batchService = new BatchService();

File: services/reporting/batch_service.dart
// lib/app/services/batch_service/batch_service.dart
import 'dart:async';
import 'package:get/get.dart';
import 'package:hse_pos/app/repos/batch_reports_repository/batch_reports_repository.dart';
import 'package:uuid/uuid.dart';
import 'package:hse_pos/app/core/locators/service_locator.dart';
import 'package:hse_pos/app/databases/hsepos_database.dart';
import 'package:hse_pos/app/models/batch_report_model/batch_report_model.dart';
import 'package:hse_pos/app/core/utils/helpers/current_location.dart';

class BatchService extends GetxService {
  final BatchReportsRepository _batchRepository;

  // Observable current active batch
  final Rxn<BatchReportModel> activeBatch = Rxn<BatchReportModel>();

  // Real-time batch stats
  final Rx<BatchReportStats?> currentBatchStats = Rx<BatchReportStats?>(null);

  // Timer for auto-closure check
  Timer? _midnightCheckTimer;

  // Stream subscriptions
  StreamSubscription? _batchWatchSubscription;

  static BatchService get instance => Get.find<BatchService>();

  BatchService({BatchReportsRepository? batchRepository})
    : _batchRepository =
          batchRepository ?? BatchReportsRepository(locator<TestDatabase>());

  @override
  Future<void> onInit() async {
    super.onInit();
    await initializeBatchSystem();
    _startMidnightCheckTimer();
  }

  @override
  void onClose() {
    _midnightCheckTimer?.cancel();
    _batchWatchSubscription?.cancel();
    super.onClose();
  }

  /// Initialize batch system on app start
  Future<void> initializeBatchSystem() async {
    try {
      final locationId = getLocationId(initial: true);
      if (locationId == null) {
        printError(info: 'Cannot initialize batch: No location ID available');
        return;
      }

      // Check for active batch
      final existingBatch = await _batchRepository.getActiveBatchReport(
        locationId: locationId.toString(),
      );

      if (existingBatch != null) {
        // Active batch exists
        activeBatch.value = existingBatch;

        // Check if it should have been closed (day changed)
        if (_shouldAutoCloseBatch(existingBatch.start_date ?? DateTime.now())) {
          await _autoCloseBatch(existingBatch);
          await _createNewBatch();
        } else {
          // Watch active batch for changes
          _watchActiveBatch(locationId.toString());
        }
      } else {
        // No active batch - create one
        await _createNewBatch();
      }
    } catch (e, stackTrace) {
      printError(info: 'Error initializing batch system: $e\n$stackTrace');
    }
  }

  /// Create a new batch report
  Future<bool> _createNewBatch() async {
    try {
      final locationId = getLocationId(initial: true);
      final terminalId = getTerminalId(initial: true);

      if (locationId == null || terminalId == null) {
        printError(
          info: 'Cannot create batch: Missing location or terminal ID',
        );
        return false;
      }

      final uuid = const Uuid().v4();
      final now = DateTime.now();
      final batchName = 'Batch ${_formatDate(now)}';

      final newBatch = BatchReportModel(
        uuid: uuid,
        batch_name: batchName,
        start_date: now,
        status: 'ACTIVE',
        location_id: locationId.toString(),
        terminal_id: terminalId.toString(),
        total_sales: 0.0,
        total_cost: 0.0,
        gross_profit: 0.0,
        net_profit: 0.0,
        total_orders: 0,
        total_items_sold: 0,
        created_at: now.toIso8601String(),
        updated_at: now.toIso8601String(),
      );

      await _batchRepository.createBatchReport(newBatch);
      activeBatch.value = newBatch;

      // Start watching the new batch
      _watchActiveBatch(locationId.toString());

      printInfo(info: 'New batch created: $batchName (UUID: $uuid)');
      return true;
    } catch (e, stackTrace) {
      printError(info: 'Error creating new batch: $e\n$stackTrace');
      return false;
    }
  }

  /// Manually start a new batch (closes current if active)
  Future<bool> startNewBatch() async {
    try {
      if (activeBatch.value != null && activeBatch.value!.status == 'ACTIVE') {
        // Close current batch first
        final closed = await closeBatch(activeBatch.value!.uuid!);
        if (!closed) {
          printError(
            info: 'Failed to close current batch before starting new one',
          );
          return false;
        }
      }

      return await _createNewBatch();
    } catch (e, stackTrace) {
      printError(info: 'Error starting new batch: $e\n$stackTrace');
      return false;
    }
  }

  /// Close the current active batch
  Future<bool> closeBatch(String uuid) async {
    try {
      // Calculate final stats
      final stats = await calculateBatchStats(uuid);

      if (stats == null) {
        printError(info: 'Cannot close batch: Failed to calculate stats');
        return false;
      }

      // Close the batch with final stats
      final success = await _batchRepository.closeBatchReport(
        uuid,
        totalSales: stats.grossSales,
        totalCost: 0.0, // Will be calculated when cost tracking is added
        grossProfit: stats.grossSales, // Temporary until cost tracking
        netProfit: stats.netSales,
        totalOrders: stats.totalOrders,
        totalItemsSold: _calculateTotalItems(stats),
        categoryBreakdown: _convertCategoryStatsToMap(stats.categoryStats),
        productBreakdown: _convertTopProductsToMap(stats.topProducts),
        paymentMethodBreakdown: stats.paymentMethodBreakdown,
        serviceTypeBreakdown: stats.serviceTypeBreakdown,
      );

      if (success) {
        activeBatch.value = activeBatch.value?.copyWith(
          status: 'CLOSED',
          end_date: DateTime.now(),
          closed_at: DateTime.now().toIso8601String(),
        );
        printInfo(info: 'Batch closed successfully: $uuid');
      }

      return success;
    } catch (e, stackTrace) {
      printError(info: 'Error closing batch: $e\n$stackTrace');
      return false;
    }
  }

  /// Auto-close batch at midnight
  Future<void> _autoCloseBatch(BatchReportModel batch) async {
    try {
      printInfo(
        info: 'Auto-closing batch due to day change: ${batch.batch_name}',
      );
      await closeBatch(batch.uuid!);
    } catch (e, stackTrace) {
      printError(info: 'Error auto-closing batch: $e\n$stackTrace');
    }
  }

  /// Check if batch should be auto-closed (day changed)
  bool _shouldAutoCloseBatch(DateTime batchStartDate) {
    final now = DateTime.now();
    return batchStartDate.year != now.year ||
        batchStartDate.month != now.month ||
        batchStartDate.day != now.day;
  }

  /// Start timer to check for midnight
  void _startMidnightCheckTimer() {
    // Calculate time until next midnight
    final now = DateTime.now();
    final tomorrow = DateTime(now.year, now.month, now.day + 1);
    final durationUntilMidnight = tomorrow.difference(now);

    // Set initial timer for midnight
    _midnightCheckTimer = Timer(durationUntilMidnight, () {
      _onMidnightReached();

      // Set up recurring daily timer
      _midnightCheckTimer = Timer.periodic(
        const Duration(days: 1),
        (_) => _onMidnightReached(),
      );
    });
  }

  /// Called when midnight is reached
  Future<void> _onMidnightReached() async {
    try {
      printInfo(info: 'Midnight reached - checking for batch auto-closure');

      if (activeBatch.value != null && activeBatch.value!.status == 'ACTIVE') {
        await _autoCloseBatch(activeBatch.value!);
        await _createNewBatch();
      }
    } catch (e, stackTrace) {
      printError(
        info: 'Error handling midnight batch closure: $e\n$stackTrace',
      );
    }
  }

  /// Watch active batch for changes
  void _watchActiveBatch(String locationId) {
    _batchWatchSubscription?.cancel();

    _batchWatchSubscription = _batchRepository
        .watchActiveBatchReport(locationId: locationId)
        .listen(
          (batch) {
            if (batch != null) {
              activeBatch.value = batch;
            }
          },
          onError: (error) {
            printError(info: 'Error watching active batch: $error');
          },
        );
  }

  /// Calculate real-time batch statistics
  Future<BatchReportStats?> calculateBatchStats(String batchUuid) async {
    try {
      final batch = await _batchRepository.getBatchReportByUuid(batchUuid);
      if (batch == null) return null;

      // Get all orders for this batch
      final database = locator<TestDatabase>();
      final ordersRepo = database.extOrdersRepository;

      final allOrders = await ordersRepo.getAllOrders(lastThreeDaysOnly: false);

      // Filter orders by batch time range
      final batchOrders = allOrders.where((order) {
        final orderDate =order.created_at ?? DateTime.now();


        final start = batch.start_date ?? DateTime.now();
        final end = batch.end_date ?? DateTime.now();

        return orderDate.isAfter(start) &&
            (batch.status == 'ACTIVE' || orderDate.isBefore(end));
      }).toList();

      // Calculate stats
      double grossSales = 0.0;
      double discounts = 0.0;
      double refunds = 0.0;
      final int totalOrders = batchOrders.length;
      int completedOrders = 0;
      int refundedOrders = 0;
      int voidedOrders = 0;
      int canceledOrders = 0;

      final Map<String, double> paymentMethodBreakdown = {};
      final Map<String, double> serviceTypeBreakdown = {};
      final Map<String, int> hourlyOrderCount = {};
      final Map<String, double> hourlySales = {};
      final Map<String, CategoryStats> categoryStats = {};
      final Map<String, StaffStats> staffStats = {};
      final Map<String, TopProductData> productDataMap = {};

      for (final order in batchOrders) {
        // Calculate order total
        double orderTotal = 0.0;
        int itemCount = 0;

        if (order.products != null) {
          for (final product in order.products!) {
            final qty = _parseDouble(product['qty'] ?? 1.0);
            final price = _parseDouble(product['selling_price'] ?? 0.0);
            orderTotal += price * qty;
            itemCount += qty.toInt();

            // Product stats
            final productUuid = product['uuid']?.toString() ?? '';
            final productName = product['name']?.toString() ?? 'Unknown';
            final categoryUuid = product['category_uuid']?.toString();

            if (productUuid.isNotEmpty) {
              if (productDataMap.containsKey(productUuid)) {
                final existing = productDataMap[productUuid]!;
                productDataMap[productUuid] = existing.copyWith(
                  quantitySold: existing.quantitySold + qty.toInt(),
                  revenue: existing.revenue + (price * qty),
                );
              } else {
                productDataMap[productUuid] = TopProductData(
                  productUuid: productUuid,
                  productName: productName,
                  quantitySold: qty.toInt(),
                  revenue: price * qty,
                  profit: price * qty, // Temporary until cost tracking
                  categoryUuid: categoryUuid,
                );
              }
            }

            // Category stats
            if (categoryUuid != null && categoryUuid.isNotEmpty) {
              if (categoryStats.containsKey(categoryUuid)) {
                final existing = categoryStats[categoryUuid]!;
                categoryStats[categoryUuid] = existing.copyWith(
                  itemsSold: existing.itemsSold + qty.toInt(),
                  revenue: existing.revenue + (price * qty),
                  orderCount: existing.orderCount + 1,
                );
              } else {
                // Get category name from database
                final categoryName = await _getCategoryName(categoryUuid);
                categoryStats[categoryUuid] = CategoryStats(
                  categoryUuid: categoryUuid,
                  categoryName: categoryName,
                  itemsSold: qty.toInt(),
                  revenue: price * qty,
                  orderCount: 1,
                  averageItemPrice: price,
                );
              }
            }
          }
        }

        // Apply discount
        if (order.discount != null && order.discount!.isNotEmpty) {
          final discountValue = _parseDiscount(order.discount, orderTotal);
          discounts += discountValue;
          orderTotal -= discountValue;
        }

        grossSales += orderTotal;

        // Order status counts
        final status = order.last_status?.toLowerCase() ?? '';
        if (status == 'completed') {
          completedOrders++;
        } else if (status == 'refund') {
          refundedOrders++;
          refunds += orderTotal;
        } else if (status == 'voided') {
          voidedOrders++;
        } else if (status == 'canceled') {
          canceledOrders++;
        }

        // Service type breakdown
        final serviceType = order.service_type ?? 'Unknown';
        serviceTypeBreakdown[serviceType] =
            (serviceTypeBreakdown[serviceType] ?? 0.0) + orderTotal;

        // Payment method breakdown
        if (order.payments != null && order.payments!.isNotEmpty) {
          for (final payment in order.payments!) {
            final method = payment['type']?.toString() ?? 'Unknown';
            final amount = _parseDouble(payment['amount'] ?? 0.0);
            paymentMethodBreakdown[method] =
                (paymentMethodBreakdown[method] ?? 0.0) + amount;
          }
        }

        // Hourly breakdown
        final orderDate = order.created_at ?? DateTime.now();
        final hour = '${orderDate.hour.toString().padLeft(2, '0')}:00';
        hourlyOrderCount[hour] = (hourlyOrderCount[hour] ?? 0) + 1;
        hourlySales[hour] = (hourlySales[hour] ?? 0.0) + orderTotal;

        // Staff stats (if staff info available in order)
        // TODO: Add staff tracking to orders
      }

      // Sort top products by revenue
      final topProducts = productDataMap.values.toList()
        ..sort((a, b) => b.revenue.compareTo(a.revenue));

      final netSales = grossSales - discounts - refunds;
      final averageOrderValue = totalOrders > 0
          ? grossSales / totalOrders
          : 0.0;

      return BatchReportStats(
        batchUuid: batchUuid,
        grossSales: grossSales,
        netSales: netSales,
        discounts: discounts,
        refunds: refunds,
        tax: 0.0, // TODO: Calculate when tax system is implemented
        totalOrders: totalOrders,
        completedOrders: completedOrders,
        refundedOrders: refundedOrders,
        voidedOrders: voidedOrders,
        canceledOrders: canceledOrders,
        averageOrderValue: averageOrderValue,
        paymentMethodBreakdown: paymentMethodBreakdown,
        serviceTypeBreakdown: serviceTypeBreakdown,
        hourlyOrderCount: hourlyOrderCount,
        hourlySales: hourlySales,
        topProducts: topProducts.take(20).toList(),
        categoryStats: categoryStats,
        staffStats: staffStats,
        calculatedAt: DateTime.now(),
      );
    } catch (e, stackTrace) {
      printError(info: 'Error calculating batch stats: $e\n$stackTrace');
      return null;
    }
  }

  /// Update batch stats in real-time
  Future<void> updateBatchStats() async {
    try {
      if (activeBatch.value == null || activeBatch.value!.status != 'ACTIVE') {
        return;
      }

      final stats = await calculateBatchStats(activeBatch.value!.uuid!);
      if (stats != null) {
        currentBatchStats.value = stats;

        // Update batch record with latest stats
        await _batchRepository.updateBatchReportStats(
          activeBatch.value!.uuid!,
          totalSales: stats.grossSales,
          totalCost: 0.0,
          grossProfit: stats.grossSales,
          netProfit: stats.netSales,
          totalOrders: stats.totalOrders,
          totalItemsSold: _calculateTotalItems(stats),
        );
      }
    } catch (e, stackTrace) {
      printError(info: 'Error updating batch stats: $e\n$stackTrace');
    }
  }

  /// Get batch history
  Future<List<BatchReportModel>> getBatchHistory({
    DateTime? startDate,
    DateTime? endDate,
    String? locationId,
  }) async {
    try {
      locationId ??= getLocationId(initial: true).toString();

      final start =
          startDate ?? DateTime.now().subtract(const Duration(days: 30));
      final end = endDate ?? DateTime.now();

      return await _batchRepository.getBatchReportsByDateRange(
        startDate: start,
        endDate: end,
        locationId: locationId,
      );
    } catch (e, stackTrace) {
      printError(info: 'Error getting batch history: $e\n$stackTrace');
      return [];
    }
  }

  /// Get specific batch by UUID
  Future<BatchReportModel?> getBatchByUuid(String uuid) async {
    try {
      return await _batchRepository.getBatchReportByUuid(uuid);
    } catch (e, stackTrace) {
      printError(info: 'Error getting batch by UUID: $e\n$stackTrace');
      return null;
    }
  }

  // Helper methods

  double _parseDouble(dynamic value) {
    if (value == null) return 0.0;
    if (value is int) return value.toDouble();
    if (value is double) return value;
    if (value is String) {
      try {
        return double.parse(value);
      } catch (_) {
        return 0.0;
      }
    }
    return 0.0;
  }

  double _parseDiscount(String? discount, double total) {
    try {
      if (discount == null || discount.isEmpty) return 0.0;

      final parts = discount.split(',');
      if (parts.length != 2) return 0.0;

      final value = double.tryParse(parts[0]) ?? 0.0;
      final type = parts[1].toLowerCase();

      if (type == 'percentage') {
        return (total * value) / 100;
      } else if (type == 'fixed') {
        return value > total ? total : value;
      }
      return 0.0;
    } catch (_) {
      return 0.0;
    }
  }

  int _calculateTotalItems(BatchReportStats stats) {
    return stats.topProducts.fold<int>(
      0,
      (sum, product) => sum + product.quantitySold,
    );
  }

  Map<String, dynamic> _convertCategoryStatsToMap(
    Map<String, CategoryStats> categoryStats,
  ) {
    return categoryStats.map(
      (key, value) => MapEntry(key, {
        'name': value.categoryName,
        'items_sold': value.itemsSold,
        'revenue': value.revenue,
        'order_count': value.orderCount,
        'average_price': value.averageItemPrice,
      }),
    );
  }

  Map<String, dynamic> _convertTopProductsToMap(List<TopProductData> products) {
    final Map<String, dynamic> result = {};
    for (var i = 0; i < products.length; i++) {
      final product = products[i];
      result[product.productUuid] = {
        'name': product.productName,
        'quantity_sold': product.quantitySold,
        'revenue': product.revenue,
        'profit': product.profit,
        'rank': i + 1,
      };
    }
    return result;
  }

  Future<String> _getCategoryName(String categoryUuid) async {
    try {
      final database = locator<TestDatabase>();
      final categoryRepo = database.categoriesRepository;
      return await categoryRepo.getCategoryNameByUuid(categoryUuid) ??
          'Unknown';
    } catch (_) {
      return 'Unknown';
    }
  }

  String _formatDate(DateTime date) {
    return '${date.year}-${date.month.toString().padLeft(2, '0')}-${date.day.toString().padLeft(2, '0')}';
  }

  void printInfo({required String info}) {
    print('[BatchService] $info');
  }

  void printError({required String info}) {
    print('[BatchService ERROR] $info');
  }

  /// Get batch statistics (for specific or current active batch)
  Future<BatchReportStats?> getBatchStats({String? batchUuid}) async {
    try {
      // Use provided UUID or the current active batch
      final targetUuid = batchUuid ?? activeBatch.value?.uuid;

      if (targetUuid == null) {
        printError(
          info:
              'Cannot get batch stats: No batch UUID provided or active batch found',
        );
        return null;
      }

      printInfo(info: 'Fetching stats for batch: $targetUuid');

      // Calculate stats based on all orders linked to this batch
      final stats = await calculateBatchStats(targetUuid);

      if (stats == null) {
        printError(info: 'Failed to calculate stats for batch: $targetUuid');
        return null;
      }

      // Store in observable (optional — keeps live data accessible)
      currentBatchStats.value = stats;

      printInfo(info: 'Batch stats fetched successfully for: $targetUuid');
      return stats;
    } catch (e, stackTrace) {
      printError(info: 'Error getting batch stats: $e\n$stackTrace');
      return null;
    }
  }
}
